src/main/scala/com/fulcrumgenomics/commons/async/AsyncRunnable.scala

```diff
@@ -0,0 +1,168 @@
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+package com.fulcrumgenomics.commons.async
+
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.atomic.{AtomicInteger, AtomicReference}
+
+import scala.concurrent.{Awaitable, ExecutionContext, Future}
+
+object AsyncRunnable {
+
+  /** A counter used for thread naming in [[AsyncRunnable]] and sub-classes */
+  private val threadsCreated = new AtomicInteger(0)
+
+  /** Generates a name for thread that will run the given runnable. */
+  private[AsyncRunnable] def nextName(asyncRunnable: AsyncRunnable): String = {
+    asyncRunnable.getClass.getSimpleName + this.threadsCreated.incrementAndGet
+  }
+}
+
+/** A trait that can be mixed in to help manage asynchronous execution while being [[Runnable]].
+  *
+  * The `execute()` method should contain the work to be performed asynchronously.
+  *
+  * The `thread()` method can be used to return a [[Thread]] that wraps this [[Runnable]].
+  *
+  * The `start()` method can be used to create and start daemon [[Thread]] that wraps this [[Runnable]].
+  *
+  * The `checkAndRaise()` method should be used by implementing classes to check if the
+  * `execute()`  method has encountered an exception, and if so, it will rethrow the exception as an
+  * [[Error]] or [[RuntimeException]].
+  *
+  * The `tryAndModifyInterruptedException()` method should wrap blocking code in implementing classes
+  * that are not part of the `execute()` method.  This method will throw a [[RuntimeException]] if interrupted while
+  * waiting.
+  */
+trait AsyncRunnable extends Runnable {
+
+  /** A reference to an exception thrown during an asynchronous operation. */
+  private[AsyncRunnable] val _throwable: AtomicReference[Throwable] = new AtomicReference(null)
+
+  /** This signals that the run method has started. */
+  private val startedLatch: CountDownLatch = new CountDownLatch(1)
+
+  /** This signals that the run method has completed, even if an exception occurred. This is needed so the close method
+    * can be sure that the run method is no longer executing. */
+  private val doneLatch: CountDownLatch = new CountDownLatch(1)
+
+  final def run(): Unit = {
+    val outer = this
```

Comment:
FYI, you never need to do this.  Here, I think you can just use `this`.  But even in the case of nested classes you can do the following:

```scala
class Foo {
    def close(): Unit = println("Foo closing!")
    class Bar {
        def close(): Unit = Foo.this.close()
    }
}
```

Reply:
Cool!